### ISSUE TITLE
[RFE] Check if `inspektor-gadget` scales and until which number of nodes

### ISSUE THREAD

**Issue by eiffel-fl** [kinvolk/inspektor-gadget#519]
## Current situation

We do not know if `inspektor-gadget` scales well.
Normally it should, as a gadget pod is deployed within each node but there is maybe a bottleneck on data gathering for each node.

## Impact

Maybe `inspektor-gadget` does not scale over X nodes, who knows?

## Ideal future situation

We should be able to know the number of nodes `inspektor-gadget` can handle.
If some minor modifications can be added to increase its scaling, they should be added.

**Comment by marga-kinvolk** [comment]
Francis, you self-assigned this bug to yourself. Do you have a specific test in mind? What are you planning to do about this issue?

**Comment by eiffel-fl** [comment]
> Do you have a specific test in mind? What are you planning to do about this issue?

Yes, the idea is to run this on AKS with a high number of node (let's call it N).
On each of the N nodes, a pod which does something like this is run:
```bash
forks=0

**Comment by alban** [comment]
We should test those IG components:
- kubeipresolver and kubenameresolver when there is a large amount of Kubernetes services, endpoints and pods and when they are added/removed regularly. We need to test gadgets that use those operators.
- trace exec with the new --cwd option
- test `ig trace exec` in addition to `kubectl-gadget trace exec` (this could be done with a `kubectl debug node` command)

It would be interesting to report the cpu usage (in percent) during the tests:
- cpu usage in userspace by the `ig` or `gadgettracermanager` process.
  - cpu profile (`sudo /usr/share/bcc/tools/profile -p $PID_OF_IG`) and report the most common stacks
- cpu usage by IG's ebpf programs (using `ig top ebpf` to measure).

**Comment by alban** [comment]
> The whole goal of testing kubectl gadget trace exec was to test the scaling on several nodes of Inspektor Gadget.
> If you want to test ig, you do not need to test it on several nodes, just on a big one (e.g. 96 cores or even more).

I was thinking that when running `ig` on a single worker node, it might still have scaling issues if there are a lot of nodes and a lot of service endpoints because the operators kubeipresolver and kubenameresolver still need to get the list of endpoints e.g. for the "trace tcp" gadget. But now I see `ig` does not yet use those operators. cc @burak-ok

> With the second, you only get the CPU usage of the eBPF program, right?
> It would be interesting to get both the eBPF program and the golang code.
> Particularly to get an understanding of at which cost we scale.

Yes exactly. Some users just want to know a global IG cpu usage percentage that includes both the eBPF programs and userspace for specific scenarios.
